newton/utils/contact_sensor.py

```diff
@@ -0,0 +1,397 @@
+        self.sensor_keys = sensor_keys
+        self.contact_partner_keys = contact_partner_keys
+        self.sensor_entities = sensor_entities
+        self.contact_partner_entities = contact_partner_entities
+        self.entity_pairs = entity_pairs
+
+        self.finalized = True
+
+
+@dataclass
+class ContactQuery:
+    """Contact Query data
+    sensor_entities: List of entity tuples representing the sensor entities.
+        These are indexed by the first element of sensor_matrix tuples.
+    select_entities: List of entity tuples representing the contact partners.
+        These are indexed by the values in the second element of the sensor_matrix tuples.
+    sensor_matrix: List of tuples defining which sensors interact with which select entities.
+        Each tuple is (sensor_index, tuple of select_indices) where:
+        - sensor_index: Index into sensor_entities list.
+        - select_indices: Tuple of indices into select_entities list.
+        The resulting contact matrix will have shape (len(sensor_matrix), max_selects)
+        where max_selects is the maximum length of select_indices tuples.
+    """
+
+    sensor_entities: list[Entity]
+    select_entities: list[Entity]
+
+    sensor_list: list[tuple[int, tuple[int, ...]]] | None = None
+    """List of sensors. Indexes sensor_entities and select_entities"""
+
+    sensor_keys: list[str] | None = None
+    select_keys: list[str] | None = None
+
+    colliding_shape_pairs: set[tuple[int, int]] | None = None
+
+
+class ContactSensorManager:
+    def __init__(self, model: Model):
+        self.contact_queries: list[ContactQuery] = []
+        self.contact_views: list[ContactView] = []
+        self.contact_reporter: ContactReporter | None = None
+        self.model: Model = model
+
+    def add_contact_query(
+        self,
+        sensor_entities: list[Entity],
+        select_entities: list[Entity | MatchAny],
+        sensor_keys: list[tuple[Any, ...]],
+        select_keys: list[tuple[Any, ...]],
+        contact_view: ContactView,
+        colliding_shape_pairs: set[tuple[int, int]] | None = None,
+    ):
+        """Add a contact query to track contact forces between specified entities.
+
+        Args:
+            query: ContactQuery object containing sensor and select indices and keys.
+            contact_view: ContactView object that will reference the results of this query.
+        """
+        query = ContactQuery(
+            sensor_entities=sensor_entities,
+            select_entities=select_entities,
+            sensor_keys=sensor_keys,
+            select_keys=select_keys,
+            colliding_shape_pairs=colliding_shape_pairs,
+        )
+
+        query.sensor_list = self._build_sensor_list(query)
+
+        self.contact_queries.append(query)
+        self.contact_views.append(contact_view)
+
+    def eval_contact_sensors(self, contact_info: ContactInfo):
+        self.contact_reporter._select_aggregate_net_force(contact_info)
+
+    def finalize(self):
```

Comment:
Same as before, maybe rename this to `evaluate()` or something more descriptive.